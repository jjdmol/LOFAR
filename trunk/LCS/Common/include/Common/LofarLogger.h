//#  LofarLogger.h: Macro interface to the lofar logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#if !defined(COMMON_LOFAR_LOGGER_H)
#define COMMON_LOFAR_LOGGER_H

#include <lofar_config.h>
#include <Common/StringUtil.h>

//# Dependent of the availability of the log4Cplus package one set of macro's
//# or the other set of macro's is included.
#ifdef HAVE_LOG4CPLUS
#include <Common/LofarLog4Cplus.h>
#else
#include <Common/LofarLogCout.h>
#endif //HAVE_LOG4CPLUS

namespace LOFAR 
{
  // Define the LOFAR::Exception class.
  EXCEPTION_CLASS(AssertError,Exception)  
}

//# --------------------- Common definitions ---------------

//# Make sure that LOFARLOGGER_PACKAGE is defined.
#ifndef LOFARLOGGER_PACKAGE
# define LOFARLOGGER_PACKAGE "Unknown_package"
#endif

//#
//# AUTO_FUNCTION_NAME
//#
// The macro AUTO_FUNCTION_NAME will be resolved by the (pre)compiler to 
// hold the name of the function the macro was used in.
#if defined(HAVE_PRETTY_FUNCTION)
#	define AUTO_FUNCTION_NAME		__PRETTY_FUNCTION__
#elif defined(HAVE_FUNCTION)
#	define AUTO_FUNCTION_NAME		__FUNCTION__
#else
#	define AUTO_FUNCTION_NAME		"??"
#endif

// NB: THROW is defined in package-dependant file.

// If the condition of assert is NOT met a logrequest is sent to the logger
// <module>.EXCEPTION and an AsserError exception is thrown.
#define ASSERT(cond) \
	if (!(cond))  THROW(LOFAR::AssertError, "Assertion: " #cond)
#define ASSERTSTR(cond,stream) \
	if (!(cond))  THROW(LOFAR::AssertError, "Assertion: " #cond "; " << stream)

// If the condition of failwhen is met a logrequest is sent to the logger
// <module>.EXCEPTION and an AsserError exception is thrown.
#define FAILWHEN(cond) \
	if (cond)  THROW(LOFAR::AssertError, "Failtest: " #cond)

// The DBG... version of ASSERT and FAILWHEN will only be in your compiled
// code when the (pre)compiler flag ENABLE_DBGASSERT is defined.
#ifdef ENABLE_DBGASSERT
#define DBGASSERT(cond)				ASSERT(cond)
#define DBGASSERTSTR(cond,stream)	ASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)			FAILWHEN(cond)
#else
#define DBGASSERT(cond)
#define DBGASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)
#endif

#endif // file read before
