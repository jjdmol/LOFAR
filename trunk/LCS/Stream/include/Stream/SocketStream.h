///# SocketStream.h: 
//#
//# Copyright (C) 2008
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

#ifndef LOFAR_LCS_STREAM_SOCKET_STREAM_H
#define LOFAR_LCS_STREAM_SOCKET_STREAM_H

#include <Common/Exception.h>
#include <Common/LofarTypes.h>
#include <Common/SystemCallException.h>
#include <Stream/FileDescriptorBasedStream.h>

#include <time.h>
#include <string>

namespace LOFAR {

class SocketStream : public FileDescriptorBasedStream
{
  public:
    EXCEPTION_CLASS(TimeOutException, LOFAR::Exception);
    SYSTEMCALLEXCEPTION_CLASS(BindException, LOFAR::SystemCallException);

    enum Protocol {
      TCP, UDP
    };

    enum Mode {
      Client, Server
    };

  	    SocketStream(const string &hostname, uint16 _port, Protocol, Mode, time_t timeout = 0, const string &nfskey = 0);
    virtual ~SocketStream();

    void    reaccept(time_t timeout = 0); // only for TCP server socket
    void    setReadBufferSize(size_t size);

    const Protocol protocol;
    const Mode mode;

  private:
    const string hostname;
    uint16 port;
    const string nfskey;
    int     listen_sk;

    void accept(time_t timeout);

    static std::string readkey(const string &nfskey, time_t &timeout);
    static void writekey(const string &nfskey, uint16 port);
    static void deletekey(const string &nfskey);
};

} // namespace LOFAR

#endif
