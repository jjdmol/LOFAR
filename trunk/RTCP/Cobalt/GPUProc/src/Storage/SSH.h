//#  SSH.h
//#
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: SSH.h 15296 2010-03-24 10:19:41Z mol $


//# Never #include <config.h> or #include <lofar_config.h> in a header file!

#ifndef LOFAR_RTCP_SSH_H
#define LOFAR_RTCP_SSH_H

#include <string>

#include <Common/Thread/Thread.h>
#include <Common/Exception.h>
#include <Stream/FileDescriptorBasedStream.h>
#include <libssh2.h>
#include <CoInterface/SmartPtr.h>
#include <time.h>
#include <string>
#include <sstream>

namespace LOFAR {
namespace RTCP {

bool SSH_Init();
void SSH_Finalize();

// Sets up and maintains an SSH connection using LibSSH2.
class SSHconnection {
public:
  // The number of seconds to wait to retry if establishing a connection fails.
  static const unsigned RETRY_DELAY = 60;

  EXCEPTION_CLASS(SSHException, LOFAR::Exception);

  SSHconnection(const std::string &logPrefix, const std::string &hostname, const std::string &commandline, const std::string &username, const std::string &pubkey, const std::string &privkey, bool captureStdout = false);

  ~SSHconnection();

  // Start connecting
  void start();

  // Abort the connection
  void cancel();

  // Wait for the connection to finish
  void wait();
  void wait( const struct timespec &deadline );

  // Returns whether the connection is finished
  bool isDone();

  // Returns whether the SSH session is (or was) connected succesfully to the
  // remote SSH daemon.
  bool connected() const;

  // If stdout is captured, return the captured output
  std::string stdoutBuffer() const;

private:
  const string itsLogPrefix;
  const string itsHostName;
  const string itsCommandLine;
  const string itsUserName;
  const string itsPublicKey;
  const string itsPrivateKey;

  bool itsConnected;

  SmartPtr<Thread> itsThread;
  const bool itsCaptureStdout;
  std::stringstream itsStdoutBuffer;

  bool waitsocket( LIBSSH2_SESSION *session, FileDescriptorBasedStream &sock );

  LIBSSH2_SESSION *open_session( FileDescriptorBasedStream &sock );
  void close_session( LIBSSH2_SESSION *session, FileDescriptorBasedStream &sock );
  LIBSSH2_CHANNEL *open_channel( LIBSSH2_SESSION *session, FileDescriptorBasedStream &sock );
  void close_channel( LIBSSH2_SESSION *session, LIBSSH2_CHANNEL *channel, FileDescriptorBasedStream &sock );

  void commThread();
};

// Discover the file name to the .ssh public/private key files,
// and put them in pubkey and privkey. Returns true if the files
// were found, and were usable for SSHconnection to localhost:22.
bool discover_ssh_keys(char *pubkey, size_t pubkey_buflen, char *privkey, size_t privkey_buflen);

} // namespace RTCP
} // namespace LOFAR


#endif
