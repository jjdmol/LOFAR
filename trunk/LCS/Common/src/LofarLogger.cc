//#  LofarLogger.cc: Interface to the log4cplus logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Includes
#include <Common/LofarLogger.h>

namespace LOFAR {

//#------------------------- Global user functions -------------------------------
//
// formatString(format, ...) --> string
//
// Define a global function the accepts printf like arguments and returns a string.
//
const std::string formatString(const	char* format, ...) {
	char		tmp_cstring[1024];
	va_list		ap;

	va_start (ap, format);
	vsnprintf(tmp_cstring, sizeof(tmp_cstring), format, ap);
	va_end   (ap);

	return   std::string(tmp_cstring);
}


//#------------------------- Internal implementation -----------------------------
using namespace log4cplus;

//# ------------------------ implement the five trace levels ------------------------
const LogLevel	TRACE1_LOG_LEVEL	= 1;
const LogLevel	TRACE2_LOG_LEVEL	= 2;
const LogLevel	TRACE3_LOG_LEVEL	= 3;
const LogLevel	TRACE4_LOG_LEVEL	= 4;
const LogLevel	TRACE5_LOG_LEVEL	= 5;

#define _TRACE1_STRING	LOG4CPLUS_TEXT("TRACE1")
#define _TRACE2_STRING	LOG4CPLUS_TEXT("TRACE2")
#define _TRACE3_STRING	LOG4CPLUS_TEXT("TRACE3")
#define _TRACE4_STRING	LOG4CPLUS_TEXT("TRACE4")
#define _TRACE5_STRING	LOG4CPLUS_TEXT("TRACE5")

tstring traceLevel2String(LogLevel ll) {
	switch (ll) {
	case TRACE1_LOG_LEVEL:	return _TRACE1_STRING;
	case TRACE2_LOG_LEVEL:	return _TRACE2_STRING;
	case TRACE3_LOG_LEVEL:	return _TRACE3_STRING;
	case TRACE4_LOG_LEVEL:	return _TRACE4_STRING;
	case TRACE5_LOG_LEVEL:	return _TRACE5_STRING;
	}

	return tstring();		// not found
}

LogLevel string2TraceLevel (const tstring& lname) {
	if (lname == _TRACE1_STRING)	return TRACE1_LOG_LEVEL;
	if (lname == _TRACE2_STRING)	return TRACE2_LOG_LEVEL;
	if (lname == _TRACE3_STRING)	return TRACE3_LOG_LEVEL;
	if (lname == _TRACE4_STRING)	return TRACE4_LOG_LEVEL;
	if (lname == _TRACE5_STRING)	return TRACE5_LOG_LEVEL;

	return NOT_SET_LOG_LEVEL;			// not found
}

// initTracemodule
//
// Function that is used when the TRACE levels are NOT compiled out. It registers
// the TRACEn management routines at the Log4Cplus LogLevelManager and sets up the
// global trace-logger named "TRACE", with the additivity set to false.
// Attached to the trace-logger is one Appender that logs to stderr.
//
void initTraceModule (void) {
	//# register our own loglevels
	getLogLevelManager().pushToStringMethod(traceLevel2String);
	getLogLevelManager().pushFromStringMethod(string2TraceLevel);
	
	//# Setup a property object to initialise the TRACE Logger
	helpers::Properties		traceProp;		
	traceProp.setProperty("log4cplus.logger.TRACE", "TRACE, TRACE");
	traceProp.setProperty("log4cplus.additivity.TRACE", "false");
	traceProp.setProperty("log4cplus.appender.TRACE", "log4cplus::ConsoleAppender");
//	traceProp.setProperty("log4cplus.appender.TRACE.layout","log4cplus::PatternLayout");
//	traceProp.setProperty("log4cplus.appender.TRACE.layout.ConversionPattern","%c %m%n");
	traceProp.setProperty("log4cplus.appender.TRACE.layout", "log4cplus::TTCCLayout");
	traceProp.setProperty("log4cplus.appender.TRACE.layout.Use_gmtime", "true");

	PropertyConfigurator(traceProp).configure();

}
LOFAR::LoggerReference	TraceLoggerRef("TRACE");			// create the tracelogger

}	// namespace LOFAR
