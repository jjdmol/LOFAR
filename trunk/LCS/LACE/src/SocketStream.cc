//#  SocketStream.cc: Abstract base class for several kinds of sockets
//#
//#  Copyright (C) 2008
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <unistd.h>					// file stuff
#include <netinet/in.h>				// IPPROTO
//#include <netdb.h>					// fileStat
#include <Common/LofarLogger.h>
#include <LACE/SocketStream.h>

namespace LOFAR {
  namespace LACE {

//
// SocketStream()
//
SocketStream::SocketStream() :
	itsIsConnected(false)
{}


//
// ~SocketStream()
//
SocketStream::~SocketStream()
{
}


//
// read (buffer, nrBytes)
//
int SocketStream::read(void*	buffer, size_t	nrBytes)
{
	if (!itsIsConnected) {
		LOG_ERROR ("SocketStream in not yet connected");
		return (-1);
	}
	return (read(buffer, nrBytes));
}

//
// write (buffer, nrBytes)
//
// Returnvalue: < 0					error occured
//				>= 0 != nrBytes		intr. occured or socket is nonblocking
//				>= 0 == nrBytes		everthing went OK.
//
int SocketStream::write(const void*	buffer, size_t	nrBytes)
{
	if (!itsIsConnected) {
		LOG_ERROR ("SocketStream in not yet connected");
		return (-1);
	}
	return (write(buffer, nrBytes));
}


  } // namespace LACE
} // namespace LOFAR
