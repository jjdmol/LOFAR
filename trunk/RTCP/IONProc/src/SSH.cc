//#  SSH.cc: setup an SSH connection using libssh2
//#
//#  Copyright (C) 2012
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: SSH.cc 18226 2012-06-09 12:56:47Z mol $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>
#include <SSH.h>

#include <Common/Thread/Cancellation.h>
#include <Common/SystemCallException.h>
#include <Common/LofarLogger.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <unistd.h>
#include <signal.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <vector>
#include <string>
#include <sstream>

#ifdef HAVE_LIBSSH2
#include <Scheduling.h>
#include <sstream>
#include <sys/select.h>
#include <Common/lofar_string.h>
#include <Stream/SocketStream.h>
#endif

using namespace std;

namespace LOFAR {
namespace RTCP {

#ifdef HAVE_LIBSSH2
  
SSHconnection::SSHconnection(const string &logPrefix, const string &hostname, const string &commandline, const string &username, const string &sshkey)
:
  itsLogPrefix(logPrefix),
  itsHostName(hostname),
  itsCommandLine(commandline),
  itsUserName(username),
  itsSSHKey(sshkey)
{
}

void SSHconnection::start()
{
  itsThread = new Thread(this, &SSHconnection::commThread, itsLogPrefix + "[SSH Thread] ", 65536);
}

void SSHconnection::stop( const struct timespec &deadline )
{
  if (!itsThread->wait(deadline)) {
    itsThread->cancel();

    itsThread->wait();
  }
}

void SSHconnection::free_session( LIBSSH2_SESSION *session )
{
  if (!session)
    return;

  libssh2_session_disconnect(session, "Normal Shutdown, Thank you for playing");
  libssh2_session_free(session);
}

void SSHconnection::free_channel( LIBSSH2_CHANNEL *channel )
{
  if (!channel)
    return;

  libssh2_channel_free(channel);
}

bool SSHconnection::open_session( FileDescriptorBasedStream &sock )
{
  int rc;

  /* Create a session instance */
  session = libssh2_session_init();
  if (!session) {
    LOG_ERROR_STR( itsLogPrefix << "Cannot create SSH session object" );
    return false;
  }

  /* tell libssh2 we want it all done non-blocking */
  libssh2_session_set_blocking(session, 0);

  /* ... start it up. This will trade welcome banners, exchange keys,
   * and setup crypto, compression, and MAC layers
   */
  while ((rc = libssh2_session_handshake(session, sock.fd)) ==
         LIBSSH2_ERROR_EAGAIN) {
    waitsocket(sock);
  }

  if (rc) {
    LOG_ERROR_STR( itsLogPrefix << "Failure establishing SSH session: " << rc);
    return false;
  }

  /* Authenticate by public key */
  while ((rc = libssh2_userauth_publickey_fromfile(session,
                      itsUserName.c_str(), // remote username
                      NULL,                // public key filename
                      itsSSHKey.c_str(),   // private key filename
                      NULL                 // password
                      )) ==
         LIBSSH2_ERROR_EAGAIN) {
    waitsocket(sock);
  }

  if (rc) {
    LOG_ERROR_STR( itsLogPrefix << "Authentication by public key failed: " << rc);
    return false;
  }

  return true;
}

bool SSHconnection::open_channel( FileDescriptorBasedStream &sock )
{
  /* Exec non-blocking on the remote host */
  while( (channel = libssh2_channel_open_session(session)) == NULL &&
         libssh2_session_last_error(session,NULL,NULL,0) ==
         LIBSSH2_ERROR_EAGAIN )
  {
    waitsocket(sock);
  }

  if (!channel)
  {
    LOG_ERROR_STR( itsLogPrefix << "Could not set up SSH channel" );
    return false;
  }

  return true;
}

bool SSHconnection::close_channel( FileDescriptorBasedStream &sock )
{
  int rc;

  while( (rc = libssh2_channel_close(channel)) == LIBSSH2_ERROR_EAGAIN ) {
    waitsocket(sock);
  }

  return true;
}

bool SSHconnection::waitsocket( FileDescriptorBasedStream &sock )
{
  struct timeval timeout;
  int rc;
  fd_set fd;
  fd_set *writefd = NULL;
  fd_set *readfd = NULL;
  int dir;

  timeout.tv_sec = 10;
  timeout.tv_usec = 0;

  FD_ZERO(&fd);

  FD_SET(sock.fd, &fd);

  /* now make sure we wait in the correct direction */
  dir = libssh2_session_block_directions(session);

  if(dir & LIBSSH2_SESSION_BLOCK_INBOUND)
      readfd = &fd;

  if(dir & LIBSSH2_SESSION_BLOCK_OUTBOUND)
      writefd = &fd;

  {
    Cancellation::enable();

    // select() is a cancellation point
    rc = ::select(sock.fd + 1, readfd, writefd, NULL, &timeout);

    Cancellation::disable();
  }   

  return rc > 0;
}

void SSHconnection::commThread()
{
#if defined HAVE_BGP_ION
  //doNotRunOnCore0();
  runOnCore0();
  //nice(19);
#endif

  SocketStream sock( itsHostName, 22, SocketStream::TCP, SocketStream::Client );

  LOG_DEBUG_STR( itsLogPrefix << "Connected" );

  int rc;
  int exitcode;
  char *exitsignal=(char *)"none";

  /* Prevent cancellation from here on -- we manually insert cancellation points to avoid
     screwing up libssh2's internal administration. */
  Cancellation::disable();
  Cancellation::point();

  if (!open_session(sock))
    return;

  if (!open_channel(sock))
    return;

  while( (rc = libssh2_channel_exec(channel, itsCommandLine.c_str())) ==
         LIBSSH2_ERROR_EAGAIN )
  {
    waitsocket(sock);
  }

  if (rc)
  {
    LOG_ERROR_STR( itsLogPrefix << "Failure starting remote command: " << rc);
    return;
  }

  LOG_DEBUG_STR( itsLogPrefix << "Remote command started, waiting for output" );

  // raw input buffer
  char data[0x1000];

  // the current line (or line remnant)
  string line("");

  /* Session I/O */
  for( ;; )
  {
    /* loop until we block */
    do {
      rc = libssh2_channel_read( channel, data, sizeof data );
      if( rc > 0 )
      {
        // create a buffer for line + data
        stringstream buffer;

        buffer << line;
        buffer.write( data, rc );

        /* extract and log lines */
        for( ;; )
        {
          Cancellation::point();

          std::getline( buffer, line );

          if (!buffer.good()) {
            // 'line' now holds the remnant
            break;
          }

          // TODO: Use logger somehow (we'd duplicate the prefix if we just use LOG_* macros..)
          cout << line << endl;
        }
      } else {
        if( rc < 0 && rc != LIBSSH2_ERROR_EAGAIN ) {
          /* no need to output this for the EAGAIN case */
          LOG_ERROR_STR( itsLogPrefix << "libssh2_channel_read returned " << rc);
        }   
      }
    } while( rc > 0 );

    /* this is due to blocking that would occur otherwise so we loop on
       this condition */
    if( rc == LIBSSH2_ERROR_EAGAIN )
    {
      waitsocket(sock);
    } else {
      /* EOF */
      break;
    }    
  }

  LOG_DEBUG_STR( itsLogPrefix << "Disconnecting" );

  close_channel(sock);

  if (rc == 0)
  {
    exitcode = libssh2_channel_get_exit_status( channel );
    libssh2_channel_get_exit_signal(channel, &exitsignal,
                                    NULL, NULL, NULL, NULL, NULL);
  } else {
    exitcode = 127;
  }

  if (exitsignal) {
    LOG_ERROR_STR(itsLogPrefix << "SSH was killed by signal " << exitsignal);
  } else if(exitcode > 0) {
    LOG_ERROR_STR(itsLogPrefix << "Exited with exit code " << exitcode << " (" << explainExitStatus(exitcode) << ")" );
  } else {
    LOG_INFO_STR(itsLogPrefix << "Terminated normally");
  }
}

#endif

static void exitwitherror( const char *errorstr )
{
  // can't cast to (void) since gcc won't allow that as a method to drop the result
  int ignoreResult;

  ignoreResult = write(STDERR_FILENO, errorstr, strlen(errorstr)+1);

  // use _exit instead of exit to avoid calling atexit handlers in both
  // the master and the child process.
  _exit(1);
}

static void execSSH(const char * const sshParams[])
{
  // DO NOT DO ANY CALL THAT GRABS A LOCK, since the lock may be held by a
  // thread that is no longer part of our address space

  // use write() for output since the Logger uses a mutex, and printf also holds locks

  // Prevent cancellation due to race conditions. A cancellation can still be pending for this JobThread, in which case one of the system calls
  // below triggers it. If this thread/process can be cancelled, there will be multiple processes running, leading to all kinds of Bad Things.
  Cancellation::disable();

  // close all file descriptors other than stdin/out/err, which might have been openend by
  // other threads at the time of fork(). We brute force over all possible fds, most of which will be invalid.
  for (int f = sysconf(_SC_OPEN_MAX); f > 2; --f)
    (void)close(f);

  // create a valid stdin from which can be read (a blocking fd created by pipe() won't suffice anymore for since at least OpenSSH 5.8)
  // rationale: this forked process inherits stdin from the parent process, which is unusable because IONProc is started in the background
  // and routed through mpirun as well. Also, it is shared by all forked processes. Nevertheless, we want Storage to be able to determine
  // when to shut down based on whether stdin is open. So we create a new stdin.
  int devzero = open("/dev/zero", O_RDONLY);

  if (devzero < 0)
    exitwitherror("cannot open /dev/zero\n");

  if (close(0) < 0)
    exitwitherror("cannot close stdin\n");

  if (dup(devzero) < 0)
    exitwitherror("cannot dup /dev/zero into stdin\n");

  if (close(devzero) < 0)
    exitwitherror("cannot close /dev/zero\n");

  if (execv("/usr/bin/ssh", const_cast<char * const *>(sshParams)) < 0)
    exitwitherror("execv failed\n");

  exitwitherror("execv succeeded but did return\n");
}


pid_t forkSSH(const std::string &logPrefix, const char *hostName, const char * const extraParams[], const char *userName, const char *sshKey)
{
  pid_t pid;

  LOG_INFO_STR(logPrefix << "Starting");

  vector<const char*> sshParams;

  const char * const defaultParams[] = {
    "ssh",
    "-q",
    "-i", sshKey,
    "-c", "blowfish",
    "-o", "StrictHostKeyChecking=no",
    "-o", "UserKnownHostsFile=/dev/null",
    "-o", "ServerAliveInterval=30",
    "-l", userName,
    hostName,
    0 };

  for( const char * const *p = defaultParams; *p != 0; p++ )
    sshParams.push_back(*p);

  for( const char * const *p = extraParams; *p != 0; p++ )
    sshParams.push_back(*p);

  sshParams.push_back(0);

  switch (pid = fork()) {
    case -1 : throw SystemCallException("fork", errno, THROW_ARGS);

    case  0 : execSSH(&sshParams[0]);
  }

  return pid;
}


void joinSSH(const std::string &logPrefix, pid_t pid, unsigned &timeout)
{
  if (pid != 0) {
    int status;

    // always try at least one waitpid(). if child has not exited, optionally
    // sleep and try again.
    for (;;) {
      pid_t ret;

      if ((ret = waitpid(pid, &status, WNOHANG)) == (pid_t)-1) {
        int error = errno;

        if (error == EINTR) {
          LOG_DEBUG_STR(logPrefix << " waitpid() was interrupted -- retrying");
          continue;
        }

        // error
        LOG_WARN_STR(logPrefix << " waitpid() failed with errno " << error);
        return;
      } else if (ret == 0) {
        // child still running
        if (timeout == 0) {
          break;
        }

        timeout--;
        sleep(1);
      } else {
        // child exited
        if (WIFSIGNALED(status) != 0)
          LOG_WARN_STR(logPrefix << "SSH was killed by signal " << WTERMSIG(status));
        else if (WEXITSTATUS(status) != 0) {

          LOG_ERROR_STR(logPrefix << " exited with exit code " << WEXITSTATUS(status) << " (" << explainExitStatus(WEXITSTATUS(status)) << ")" );
        } else
          LOG_INFO_STR(logPrefix << " terminated normally");

        return;  
      }
    }

    // child did not exit within the given timeout

    LOG_WARN_STR(logPrefix << " sending SIGTERM");
    kill(pid, SIGTERM);

    if (waitpid(pid, &status, 0) == -1) {
      LOG_WARN_STR(logPrefix << " waitpid() failed");
    }

    LOG_WARN_STR(logPrefix << " terminated after sending SIGTERM");
  }
}


const char *explainExitStatus( int exitstatus )
{
  const char *explanation;

  switch (exitstatus) {
    default:
      explanation = "??";
      break;

    case 255:
      explanation = "Network or authentication error";
      break;
    case 127:
      explanation = "BASH: command/library not found";
      break;
    case 126:
      explanation = "BASH: command found but could not be executed (wrong architecture?)";
      break;

    case 128 + SIGHUP:
      explanation = "killed by SIGHUP";
      break;
    case 128 + SIGINT:
      explanation = "killed by SIGINT (Ctrl-C)";
      break;
    case 128 + SIGQUIT:
      explanation = "killed by SIGQUIT";
      break;
    case 128 + SIGILL:
      explanation = "illegal instruction";
      break;
    case 128 + SIGABRT:
      explanation = "killed by SIGABRT";
      break;
    case 128 + SIGKILL:
      explanation = "killed by SIGKILL";
      break;
    case 128 + SIGSEGV:
      explanation = "segmentation fault";
      break;
    case 128 + SIGPIPE:
      explanation = "broken pipe";
      break;
    case 128 + SIGALRM:
      explanation = "killed by SIGALRM";
      break;
    case 128 + SIGTERM:
      explanation = "killed by SIGTERM";
      break;
  }

  return explanation;
}

} // namespace RTCP
} // namespace LOFAR

