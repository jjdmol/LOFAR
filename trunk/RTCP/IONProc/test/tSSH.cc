#include <lofar_config.h>

#include <SSH.h>
#include <unistd.h>
#include <time.h>
#include <cstdlib>
#include <cstdio>
#include <Stream/SocketStream.h>
#include <Common/LofarLogger.h>

// some useful environment variables
char *USER;
char *HOME;

// the existence of $HOME/.ssh/id_rsa is assumed,
// as well as the fact that it can be used to
// authenticate on localhost.
char privkey[1024];

using namespace LOFAR;
using namespace RTCP;


void test_SSHconnection() {
#ifdef HAVE_LIBSSH2
  SSHconnection ssh("", "localhost", "echo SSHconnection success", USER, privkey);

  ssh.start();

  struct timespec ts;
  ts.tv_sec = time(0) + 10;
  ts.tv_nsec = 0;

  ssh.stop(ts);
#endif
}


void test_forkExec() {
  pid_t pid;

  char * const params[] = {
    "echo",
    "forkExec success",
    0
  };
  
  pid = forkSSH("", "localhost", params, USER, privkey);

  unsigned timeout = 10;
  joinSSH("", pid, timeout);
}

int main() {
  INIT_LOGGER( "tSSH" );

  USER = getenv("USER");
  HOME = getenv("HOME");
  snprintf(privkey, sizeof privkey, "%s/.ssh/id_rsa", HOME);

  test_SSHconnection();
  test_forkExec();

  return 0;
}
