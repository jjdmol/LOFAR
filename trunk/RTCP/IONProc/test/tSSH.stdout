system success
SSHconnection success
forkExec success
